File: agent-runner/src/rules.ts
/**
 * Permission rule evaluation for tool approval.
 *
 * Rules follow Claude Code's convention:
 *   - Format: Tool or Tool(specifier)
 *   - Evaluation order: deny → ask → allow (first match wins)
 *   - Specifiers support glob wildcards (*)
 */

export interface PermissionRule {
  id: string;
  tool: string; // "Bash", "Read", "Write", "Edit", etc.
  specifier: string; // "npm run *", "./.env", "domain:example.com", "" for match-all
  action: "allow" | "deny" | "ask";
  scope: "user" | "workspace";
  workspaceId?: string;
}

/**
 * Build a specifier string from a tool's input, used for rule matching.
 * Returns null if no meaningful specifier can be extracted.
 */
export function buildSpecifier(
  toolName: string,
  toolInput: Record<string, unknown>,
): string | null {
  switch (toolName) {
    case "Bash":
      return (toolInput.command as string) || null;
    case "Read":
    case "Write":
    case "Edit":
    case "NotebookEdit":
      return (toolInput.file_path as string) || null;
    case "WebFetch": {
      const url = toolInput.url as string;
      if (!url) return null;
      try {
        return `domain:${new URL(url).hostname}`;
      } catch {
        return null;
      }
    }
    case "Glob":
      return (toolInput.pattern as string) || null;
    case "Grep":
      return (toolInput.pattern as string) || null;
    default:
      // MCP tools: mcp__server__tool — use tool name as specifier
      if (toolName.startsWith("mcp__")) return toolName;
      return null;
  }
}

/**
 * Evaluate rules against a tool invocation.
 * Returns the matching action or null if no rule matches.
 * Evaluation order: deny → ask → allow (first match wins within each tier).
 */
export function evaluateRules(
  rules: PermissionRule[],
  toolName: string,
  specifier: string | null,
): "allow" | "deny" | "ask" | null {
  // Phase 1: deny rules
  for (const rule of rules) {
    if (rule.action === "deny" && matchesRule(rule, toolName, specifier)) {
      return "deny";
    }
  }
  // Phase 2: ask rules
  for (const rule of rules) {
    if (rule.action === "ask" && matchesRule(rule, toolName, specifier)) {
      return "ask";
    }
  }
  // Phase 3: allow rules
  for (const rule of rules) {
    if (rule.action === "allow" && matchesRule(rule, toolName, specifier)) {
      return "allow";
    }
  }
  return null;
}

/**
 * Check if a single rule matches the given tool name and specifier.
 */
function matchesRule(
  rule: PermissionRule,
  toolName: string,
  specifier: string | null,
): boolean {
  if (rule.tool !== toolName) return false;
  // Bare tool name (no specifier) matches all uses of the tool
  if (!rule.specifier) return true;
  // If rule has a specifier but the tool call has none, no match
  if (!specifier) return false;
  return wildcardMatch(rule.specifier, specifier);
}

/**
 * Glob-style wildcard matching (iterative, no regex — immune to ReDoS).
 * Supports `*` as a wildcard that matches any sequence of characters.
 *
 * Following Claude Code conventions:
 *   - `Bash(ls *)` matches "ls -la" but not "lsof" (space before * = word boundary)
 *   - `Bash(ls*)` matches both "ls -la" and "lsof"
 *   - Multiple wildcards are supported: `git * main` matches "git checkout main"
 *
 * Uses a two-pointer greedy algorithm with backtracking anchored at the last `*`.
 */
export function wildcardMatch(pattern: string, value: string): boolean {
  let pi = 0; // pattern index
  let vi = 0; // value index
  let starPi = -1; // last * position in pattern
  let matchVi = -1; // value index when last * was seen

  while (vi < value.length) {
    if (pi < pattern.length && pattern[pi] === "*") {
      // Record * position and advance pattern
      starPi = pi;
      matchVi = vi;
      pi++;
    } else if (pi < pattern.length && pattern[pi] === value[vi]) {
      // Characters match — advance both
      pi++;
      vi++;
    } else if (starPi !== -1) {
      // Mismatch but we have a prior * — backtrack:
      // let * consume one more character from value
      pi = starPi + 1;
      matchVi++;
      vi = matchVi;
    } else {
      // Mismatch with no * to fall back on
      return false;
    }
  }

  // Consume any trailing *s in the pattern
  while (pi < pattern.length && pattern[pi] === "*") {
    pi++;
  }

  return pi === pattern.length;
}

File: backend/permission/rules.go
package permission

import (
	"encoding/json"
	"os"
	"strings"
)

// Rule represents a persistent permission rule loaded from configuration.
type Rule struct {
	Tool      string `json:"tool"`               // Tool name (e.g., "Bash", "Write")
	Specifier string `json:"specifier,omitempty"` // Optional content pattern (e.g., "npm run *")
	Content   string `json:"content,omitempty"`   // Optional substring match on tool input content
	Action    string `json:"action"`              // "allow", "deny", or "ask"
}

// RuleSet holds persistent rules and provides evaluation.
type RuleSet struct {
	rules []Rule
}

// NewRuleSet creates a RuleSet from a slice of rules.
func NewRuleSet(rules []Rule) *RuleSet {
	if rules == nil {
		rules = []Rule{}
	}
	return &RuleSet{rules: rules}
}

// LoadRulesFromFile loads rules from a JSON file.
// Returns an empty RuleSet (no error) if the file does not exist.
func LoadRulesFromFile(path string) (*RuleSet, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return NewRuleSet(nil), nil
		}
		return nil, err
	}

	var rules []Rule
	if err := json.Unmarshal(data, &rules); err != nil {
		return nil, err
	}

	return NewRuleSet(rules), nil
}

// Evaluate checks all rules against a tool invocation.
// Evaluation order: deny rules first, then ask rules, then allow rules.
// First match within each tier wins. Returns "allow", "deny", "ask", or "" if no match.
func (rs *RuleSet) Evaluate(toolName, specifier string) string {
	// Phase 1: Check deny rules
	for _, r := range rs.rules {
		if r.Action == "deny" && matchesRule(r, toolName, specifier) {
			return "deny"
		}
	}

	// Phase 2: Check ask rules
	for _, r := range rs.rules {
		if r.Action == "ask" && matchesRule(r, toolName, specifier) {
			return "ask"
		}
	}

	// Phase 3: Check allow rules
	for _, r := range rs.rules {
		if r.Action == "allow" && matchesRule(r, toolName, specifier) {
			return "allow"
		}
	}

	return ""
}

// Count returns the number of rules in the set.
func (rs *RuleSet) Count() int {
	return len(rs.rules)
}

// matchesRule checks whether a rule matches a given tool name and specifier.
func matchesRule(rule Rule, toolName, specifier string) bool {
	// Tool name must match exactly
	if rule.Tool != toolName {
		return false
	}

	// If the rule has a specifier, the specifier must match
	if rule.Specifier != "" {
		if specifier == "" {
			return false
		}
		if !WildcardMatch(rule.Specifier, specifier) {
			return false
		}
	}

	// If the rule has a content pattern, the specifier must contain it as a substring
	if rule.Content != "" {
		if specifier == "" {
			return false
		}
		if !strings.Contains(specifier, rule.Content) {
			return false
		}
	}

	return true
}

// WildcardMatch performs glob-style matching with * wildcards.
// Uses a two-pointer backtracking algorithm (ReDoS-safe, O(n*m) worst case).
// A space before * acts as a word boundary — it prevents matching within a word.
// Ported from agent-runner/src/rules.ts wildcardMatch().
func WildcardMatch(pattern, value string) bool {
	// Exact match fast path
	if pattern == value {
		return true
	}

	// No wildcards — must be exact
	if !strings.Contains(pattern, "*") {
		return pattern == value
	}

	pi := 0 // pattern index
	vi := 0 // value index
	starPI := -1
	starVI := -1

	for vi < len(value) {
		if pi < len(pattern) && pattern[pi] == '*' {
			// Record star position for backtracking
			starPI = pi
			starVI = vi
			pi++
			continue
		}

		if pi < len(pattern) && (pattern[pi] == value[vi] || pattern[pi] == '?') {
			pi++
			vi++
			continue
		}

		// Mismatch — backtrack to last star
		if starPI >= 0 {
			// Check word boundary: if there's a space before the star in the pattern,
			// the star should not match within a word (space acts as boundary)
			pi = starPI + 1
			starVI++
			vi = starVI
			continue
		}

		// No star to backtrack to — mismatch
		return false
	}

	// Consume trailing stars in pattern
	for pi < len(pattern) && pattern[pi] == '*' {
		pi++
	}

	return pi == len(pattern)
}
